RecommenderSystems/dcn/dcn_train_eval.py

```diff
@@ -0,0 +1,832 @@
+    rank = flow.env.get_rank()
+    if rank == 0:
+        labels = np.concatenate(labels, axis=0)
+        preds = np.concatenate(preds, axis=0)
+        eval_time = time.time() - eval_start_time
+        auc_start_time = time.time()
+        auc = flow.roc_auc_score(labels, preds).numpy()[0]
+        auc_time = time.time() - auc_start_time
+        log_loss_start_time = time.time()
+        logloss = log_loss(labels, preds, eps=1e-7)
+        log_loss_time = time.time() - log_loss_start_time
+
+        host_mem_mb = psutil.Process().memory_info().rss // (1024 * 1024)
+        stream = os.popen("nvidia-smi --query-gpu=memory.used --format=csv")
+        device_mem_str = stream.read().split("\n")[rank + 1]
+
+        strtime = time.strftime("%Y-%m-%d %H:%M:%S")
+        print("=== test ===")
+        print(
+            f"Rank[{rank}], Epoch {epoch}, Step {cur_step}, AUC {auc:0.6f}, LogLoss {logloss:0.6f}, Eval_time {eval_time:0.2f} s, "
+            + f"AUC_time {auc_time:0.2f} s, LogLoss_time {log_loss_time: 0.2f} s, Eval_samples {labels.shape[0]}, "
+            + f"GPU_Memory {device_mem_str}, Host_Memory {host_mem_mb} MiB, {strtime}"
+        )
+
+    flow.comm.barrier()
+    return auc, logloss
+
+
+def plot_train_curve(args, train_losses):
+    plt.plot(range(1, len(train_losses) + 1), train_losses)
+    plt.ylabel("Training Logloss")
+    plt.xlabel(f"batch number (per {args.loss_print_interval} batches)")
+    plt.savefig("of_training_curve.png")
+
+
+def load_fuxi_test(args):
+
+    rank = flow.env.get_rank()
+    dcn_module = make_dcn_module(args)
+    dcn_module.to_global(flow.env.all_device_placement("cuda"), flow.sbp.broadcast)
+    state_dict = flow.load(
+        "your_path/fuxi_2_of_state_dict", global_src_rank=0
+    )
+    dcn_module.load_state_dict(state_dict, strict=False)
+
+    # dcn_module.load_state_dict(state_dict, strict=False)
+    # dcn_module.to_global(flow.env.all_device_placement("cuda"), flow.sbp.broadcast)
+    eval_graph = DCNValGraph(dcn_module, args.amp)
+    auc, logloss = eval_test(args, eval_graph, 0, 0)
+    print("test finished")
+    print("test auc : ", auc, "logloss: ", logloss)
+
+
+def get_table_size_array(feature_map_json):
+    print("Load feature_map from json: " + feature_map_json)
+    with open(feature_map_json, "r", encoding="utf-8") as fd:
+        feature_map = json.load(fd, object_pairs_hook=OrderedDict)
+    feature_specs = OrderedDict(feature_map["feature_specs"])
+    sparse_names = [f"C{i}" for i in range(1, 27)]
+    dense_names = [f"I{i}" for i in range(1, 14)]
+    columns = dense_names + sparse_names
+    table_size_array = []
+    for col in columns:
+        table_size_array.append(feature_specs[col]["vocab_size"])
+
+    return table_size_array
+
+
+if __name__ == "__main__":
+    os.system(sys.executable + " -m oneflow --doctor")
+    flow.boxing.nccl.enable_all_to_all(True)
+    args = get_args()
+    args.table_size_array = get_table_size_array(args.feature_map_json)
```

Comment:
table_size_array is overwrited

Reply:
It's a good idea to use "feature_map.json", but currently I suggest to use table_size_array only.  We consider using it in the future.